File: medium/2265-count-nodes-equal-to-average-of-subtree/solution.py
from typing import Optional, Tuple


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def averageOfSubtree(self, root: Optional["TreeNode"]) -> int:
        """
        2265. Count Nodes Equal to Average of Subtree
        Time: O(n)
        Space: O(h)
        """
        self.answer = 0

        def dfs(node: Optional["TreeNode"]) -> Tuple[int, int]:
            # returns (sum of subtree values, number of nodes)
            if node is None:
                return 0, 0

            left_sum, left_count = dfs(node.left)
            right_sum, right_count = dfs(node.right)

            total = left_sum + right_sum + node.val
            count = left_count + right_count + 1

            # values are non-negative, so // is the required floor
            if node.val == total // count:
                self.answer += 1

            return total, count

        dfs(root)
        return self.answer

File: medium/2265-count-nodes-equal-to-average-of-subtree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
 * 2265. Count Nodes Equal to Average of Subtree
 * Time: O(n)
 * Space: O(h)
 */
class Solution {
    private int answer = 0;

    public int averageOfSubtree(TreeNode root) {
        dfs(root);
        return answer;
    }

    /** Returns {sum of subtree values, number of nodes}. */
    private int[] dfs(TreeNode node) {
        if (node == null) {
            return new int[]{0, 0};
        }

        int[] left = dfs(node.left);
        int[] right = dfs(node.right);

        int total = left[0] + right[0] + node.val;
        int count = left[1] + right[1] + 1;

        // values are non-negative, so integer division is the required floor
        if (node.val == total / count) {
            answer++;
        }

        return new int[]{total, count};
    }
}
